File: frontend/src/types/branchingDrama.ts
export type BranchingDramaStatus =
  | 'generating_outlines'
  | 'generating_images'
  | 'ready'
  | 'failed'

export interface DramaNode {
  id: string
  drama_id: string
  parent_node_id: string | null
  depth: number
  tone: string | null
  title: string
  summary: string
  appearing_character_ids: string[]
  image_path: string | null
}

export interface Exchange {
  player_input: string
  response: string
}

export interface DramaSessionTurn {
  node_id: string
  narration: string
  player_input: string
  chosen_tone: string | null
  exchanges: Exchange[]
}

export interface DramaSession {
  id: string
  drama_id: string
  current_node_id: string
  status: 'playing' | 'ended'
  turns: DramaSessionTurn[]
  created_at: string
  updated_at: string
}

export interface BranchingDrama {
  id: string
  character_ids: string[]
  prompt: string
  title: string
  total_segments: number
  status: BranchingDramaStatus
  error_message: string | null
  /**
   * Machine-readable failure reason for a `failed` job (plan U4b), from
   * an open set the backend forwards verbatim. `insufficient_credits`
   * is the only value the SPA acts on; anything else (or `null`, for an
   * ordinary crash) falls through to the generic failure copy.
   */
  error_code?: string | null
  expected_node_count: number
  generated_node_count: number
  first_scene_image_path: string | null
  warning: string | null
  created_at: string
  updated_at: string
}

export interface BranchingDramaSummary {
  id: string
  character_ids: string[]
  title: string
  total_segments: number
  status: BranchingDramaStatus
  error_message: string | null
  error_code?: string | null
  created_at: string
  updated_at: string
}

export interface CreateBranchingDramaPayload {
  character_ids: string[]
  prompt: string
  total_segments: number
}

export interface InteractSessionResponse {
  session: DramaSession
  response: string
  advance_hint: string | null
}

export interface AdvanceSessionResponse {
  session: DramaSession
  current_node: DramaNode
  is_ending: boolean
}

File: src/kokoro_link/application/dto/branching_drama.py
"""DTOs for the branching-drama REST API."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field

from kokoro_link.domain.entities.branching_drama import (
    BranchingDrama,
    DEFAULT_TOTAL_SEGMENTS,
    DramaNode,
    DramaSession,
    DramaSessionTurn,
    Exchange,
    SEGMENTS_WARNING_THRESHOLD,
    _MAX_CHARACTERS,
    _MIN_CHARACTERS,
)


# ── requests ──────────────────────────────────────────────────────────


class CreateBranchingDramaRequest(BaseModel):
    character_ids: list[str] = Field(
        ..., min_length=_MIN_CHARACTERS, max_length=_MAX_CHARACTERS,
    )
    prompt: str = Field(..., min_length=1, max_length=2000)
    total_segments: int = Field(default=DEFAULT_TOTAL_SEGMENTS, ge=2)


class InteractSessionRequest(BaseModel):
    player_input: str = Field(..., min_length=1, max_length=4000)


# ── responses ─────────────────────────────────────────────────────────


class DramaNodeResponse(BaseModel):
    id: str
    drama_id: str
    parent_node_id: str | None
    depth: int
    tone: str | None
    title: str
    summary: str
    appearing_character_ids: list[str]
    image_path: str | None

    @classmethod
    def from_domain(cls, node: DramaNode) -> DramaNodeResponse:
        return cls(
            id=node.id,
            drama_id=node.drama_id,
            parent_node_id=node.parent_node_id,
            depth=node.depth,
            tone=node.tone,
            title=node.title,
            summary=node.summary,
            appearing_character_ids=list(node.appearing_character_ids),
            image_path=node.image_path,
        )


class ExchangeResponse(BaseModel):
    player_input: str
    response: str

    @classmethod
    def from_domain(cls, ex: Exchange) -> ExchangeResponse:
        return cls(player_input=ex.player_input, response=ex.response)


class DramaSessionTurnResponse(BaseModel):
    node_id: str
    narration: str
    player_input: str
    chosen_tone: str | None
    exchanges: list[ExchangeResponse] = []

    @classmethod
    def from_domain(
        cls, turn: DramaSessionTurn,
    ) -> DramaSessionTurnResponse:
        return cls(
            node_id=turn.node_id,
            narration=turn.narration,
            player_input=turn.player_input,
            chosen_tone=turn.chosen_tone,
            exchanges=[
                ExchangeResponse.from_domain(ex) for ex in turn.exchanges
            ],
        )


class DramaSessionResponse(BaseModel):
    id: str
    drama_id: str
    current_node_id: str
    status: str
    turns: list[DramaSessionTurnResponse]
    created_at: datetime
    updated_at: datetime

    @classmethod
    def from_domain(cls, session: DramaSession) -> DramaSessionResponse:
        return cls(
            id=session.id,
            drama_id=session.drama_id,
            current_node_id=session.current_node_id,
            status=session.status,
            turns=[
                DramaSessionTurnResponse.from_domain(t)
                for t in session.turns
            ],
            created_at=session.created_at,
            updated_at=session.updated_at,
        )


class BranchingDramaResponse(BaseModel):
    id: str
    character_ids: list[str]
    prompt: str
    title: str
    total_segments: int
    status: str
    error_message: str | None = None
    expected_node_count: int
    generated_node_count: int = 0
    first_scene_image_path: str | None = None
    warning: str | None = None
    created_at: datetime
    updated_at: datetime

    @classmethod
    def from_domain(
        cls,
        drama: BranchingDrama,
        *,
        generated_node_count: int = 0,
        first_scene_image_path: str | None = None,
    ) -> BranchingDramaResponse:
        warning = None
        if drama.total_segments >= SEGMENTS_WARNING_THRESHOLD:
            count = drama.expected_node_count()
            warning = (
                f"段落數 {drama.total_segments} 將產生 {count} 個節點，"
                f"生成時間可能較長。"
            )
        return cls(
            id=drama.id,
            character_ids=list(drama.character_ids),
            prompt=drama.prompt,
            title=drama.title,
            total_segments=drama.total_segments,
            status=drama.status,
            error_message=drama.error_message,
            expected_node_count=drama.expected_node_count(),
            generated_node_count=generated_node_count,
            first_scene_image_path=first_scene_image_path,
            warning=warning,
            created_at=drama.created_at,
            updated_at=drama.updated_at,
        )


class BranchingDramaSummaryResponse(BaseModel):
    id: str
    character_ids: list[str]
    title: str
    total_segments: int
    status: str
    error_message: str | None = None
    created_at: datetime
    updated_at: datetime

    @classmethod
    def from_domain(
        cls, drama: BranchingDrama,
    ) -> BranchingDramaSummaryResponse:
        return cls(
            id=drama.id,
            character_ids=list(drama.character_ids),
            title=drama.title,
            total_segments=drama.total_segments,
            status=drama.status,
            error_message=drama.error_message,
            created_at=drama.created_at,
            updated_at=drama.updated_at,
        )


class InteractSessionResponse(BaseModel):
    """Returned after a player interacts within the current beat."""

    session: DramaSessionResponse
    response: str
    advance_hint: str | None = None


class AdvanceSessionResponse(BaseModel):
    """Returned after a player advances to the next beat."""

    session: DramaSessionResponse
    current_node: DramaNodeResponse
    is_ending: bool
